test/shmpreopen.c

```diff
@@ -0,0 +1,234 @@
+
+
+/*
+ * Copyright (c) 2016 Jonathan Anderson & Stanley Uche Godfrey
+ * All rights reserved.
+ *
+ * This software was developed at Memorial University under the
+ * NSERC Discovery program (RGPIN-2015-06048).
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+/*
+ * RUN: %cc -c %cflags -D TEST_DATA_DIR="\"%p/Inputs\"" %s -o %t.o
+ * RUN: %cc %t.o %ldflags -o %t
+ * RUN: %t > %t.out
+ * RUN: %filecheck %s -input-file %t.out
+ */
+
+#include <sys/param.h>
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <sys/wait.h>
+#include "libpreopen.h"
+#include<ctype.h>
+#include <sys/capsicum.h>
+#include<err.h>
+#include<errno.h>
+#include<sys/types.h>
+#include<sys/socket.h>
+#include<sys/un.h>
+
+
+#define TEST_DIR(name) \
+	"/" TEST_DATA_DIR name
+
+static int
+send_file_descriptor(
+  int socket, /* Socket through which the file descriptor is passed */
+  int fd_to_send); /* File descriptor to be passed, could be another socket */
+
+static int
+recv_file_descriptor(
+  int socket); /* Socket from which the file descriptor is read */
+
+static void print_map_content( struct po_map *map);
+
+int main(int argc, char *argv[])
+{	pid_t child[2]; int k;
+	int sockets[2];
+	 if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) < 0) {
+     		 perror("opening stream socket pair");
+      		exit(1);
+   	}
+	//Creating first child process
```

Comment:
Why create two child processes? Forking once should be sufficient: the logic would then be, `if (child == 0) { /* I'm the child */ } else { /* I'm the parent */ }`.